packages/lib/gpcircuits/src/list-membership.ts

```diff
@@ -0,0 +1,131 @@
+   * list membership checks (via `ListMembershipModule`) of each of the `N`
+   * value hashes in the corresponding list by forming appropriate lower
+   * order tuples and one combined membership list.
+   *
+   * @param `listIndices` denotes the indices of the POD entry values to be checked
+   * @param `membershipLists` is a list each of whose elements is a membership
+   *  list for the value associated with the corresponding index in `listIndices`.
+   * @param `paramMaxValues` denotes the maximum number of POD entry values in the
+   * circuit.
+   * @param `paramTupleArity` denotes the fixed tuple arity for the circuit.
+   * @param `paramMaxListEntries` denotes the maximum number of list entries in the circuit.
+   * @param `paramMaxTuples` denotes the maximum number of tuples in the circuit.
+   * @returns `number` containing the index of the hash whose membership
+   *  is to be checked, where an index less than `paramMaxValues` corresponds to
+   *  an entry value hash and a greater index corresponds to a tuple index.
+   * @returns `number[][]` containing `paramMaxTuples` lists of `paramTupleArity` tuples
+   *  of indices, where an index `i` refers to an entry value hash if it is less than
+   *  `paramMaxValues`, otherwise it refers to the tuple represented by the
+   *  `paramMaxValues - i`th entry of this list.
+   * @returns `bigint[]` containing the combined membership list.
+   * @throws TypeError if the lengths of `listIndices` and `membershipLists` don't agree.
+   * @throws TypeError if the list of all combinations of elements of `membershipLists`
+   *  exceeds `paramMaxListentries`.
+   * @throws TypeError if the tuple arity does not lie in the acceptable range.
+   * @throws TypeError if any element of `listIndices` exceeds `paramMaxValues - 1`.
+   * @throws TypeError if the number of tuples of arity `paramTupleArity` required to
+   *  represent `listIndices` exceeds `paramMaxTuples`.
+   */
+  public static generateSignals(listIndices: number[],
+				membershipLists: PODValue[][],
+				paramMaxValues: number,
+				paramTupleArity: number,
+				paramMaxListEntries: number,
+				paramMaxTuples: number
+			       ): [number, bigint[][], bigint[]] {
+    // Check input
+    if (listIndices.length != membershipLists.length)
+      throw new TypeError("The lengths of the index and value list arrays must be the same.");
+    if (membershipLists.reduce((prod, lis) => prod*lis.length, 0) > paramMaxListEntries)
+      throw new TypeError(`The total number of entries in the list membership check exceeds MAX_LIST_ENTRIES (${paramMaxListEntries}).`);
+    if ((paramTupleArity < 2) || (paramTupleArity > 4))
+      throw new TypeError("The arity must lie between 2 and 4.");
+    listIndices.forEach(i => {if(i >= paramMaxValues) throw new Error("The indices cannot exceed MAX_VALUES.");});
+
+    // Helpers
+    const pad = (arr: any, len: number, padding: any) => arr.concat(Array(len - arr.length).fill(padding));
```

Comment:
Do these need to be `any`?  I think our lint rules will complain about this.  I certainly prefer to avoid uses of `any`, which turn off most type checking.